src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/BalloonPopup.java

```diff
@@ -52,56 +79,144 @@ public boolean isVisible() {
+    final int availableSizeLeft = componentWindowPos.x;
+    final int availableSizeRight = windowSize.width - (componentWindowPos.x + componentSize.width);
+    final int availableSizeTop = componentWindowPos.y;
+    final int availableSizeBottom = windowSize.height - (componentWindowPos.y + componentSize.height);
 
-    int mostHorizontalSpace = Integer.max(availableSizeLeft, availableSizeRight);
-    int mostVerticalSpace = Integer.max(availableSizeTop, availableSizeBottom);
+    final int mostHorizontalSpace = Integer.max(availableSizeLeft, availableSizeRight);
+    final int mostVerticalSpace = Integer.max(availableSizeTop, availableSizeBottom);
 
-    boolean positionHorizontally = mostHorizontalSpace > mostVerticalSpace;
+    final boolean canPlaceHorizontally = mostHorizontalSpace > popupWidth + 2 * POPUP_MARGIN;
+    final boolean canPlaceVertically = mostVerticalSpace > popupHeight + 2 * POPUP_MARGIN;
 
     int popupX;
     int popupY;
 
-    if (positionHorizontally) {
-      if (availableSizeRight > availableSizeLeft) {
-        popupX = componentWindowPos.x + anchorComponent.getWidth() + 5;
-      } else {
-        popupX = componentWindowPos.x - popupWidth - 5;
-      }
+    if (!canPlaceHorizontally && !canPlaceVertically) {
+      // if there's no space on any side of the component, place the popup on the component
+      popupX = componentWindowPos.x + componentSize.width - popupWidth - POPUP_MARGIN;
+      popupY = componentWindowPos.y + POPUP_MARGIN;
+
+      final var popupBounds = new Rectangle(0, 0, popupWidth, popupHeight);
+      final var mousePos = getMousePosition();
 
-      popupY = componentWindowPos.y + (anchorComponent.getHeight() - popupHeight) / 2;
+      transparencyHandler.update(mousePos != null && popupBounds.contains(mousePos));
+      if (transparencyHandler.isInAnimation()) {
+        revalidate();
+        repaint();
+      }
     } else {
-      if (availableSizeBottom > availableSizeTop) {
-        popupY = componentWindowPos.y + anchorComponent.getHeight() + 5;
+      if (!canPlaceVertically || (mostHorizontalSpace > mostVerticalSpace && canPlaceHorizontally)) {
+        if (availableSizeRight > availableSizeLeft) {
+          popupX = componentWindowPos.x + anchorComponent.getWidth() + 5;
+        } else {
+          popupX = componentWindowPos.x - popupWidth - 5;
+        }
+
+        popupY = componentWindowPos.y + (anchorComponent.getHeight() - popupHeight) / 2;
       } else {
-        popupY = componentWindowPos.y - popupHeight - 5;
+        if (availableSizeBottom > availableSizeTop) {
+          popupY = componentWindowPos.y + anchorComponent.getHeight() + 5;
+        } else {
+          popupY = componentWindowPos.y - popupHeight - 5;
+        }
+
+        popupX = componentWindowPos.x + (anchorComponent.getWidth() - popupWidth) / 2;
       }
 
-      popupX = componentWindowPos.x + (anchorComponent.getWidth() - popupWidth) / 2;
+      setTransparencyCoefficient(1.0f);
```

Comment:
Hmm... why?

Reply:
This is when the popup is placed near a component's side (not on the component) and in that case it's never transparent - this resets transparency level to fully opaque